### DISCUSSION TITLE
Micropython - simple html formular field for data input and output

### DISCUSSION THREAD

**Discussion by danieldownload** [miguelgrinberg/microdot#119] [Q&A] [answered]
Hi there,

I'm trying to migrate a project from Arduino to Micropython. I'm using an esp32 dev board.

Can someone post a simple example of a http text field which can be filled with date, submit sends the date to the backend (a file in my case).
When I open the page next time I want to load the previous data back into the text field.

Getting the data via submit button is not the problem, but I cannot get it to work to fill a text field when I open the site.

Any help would be great.

**Comment by miguelgrinberg** [comment]
This is completely unrelated to Microdot, but in any case, you can pass a `value` attribute to your text field to include the data that you want the field to be prepopulated with.

```html
<input name="my_field" type="text" value="This will appear in the field!">
```

**Reply by danieldownload** [reply]
The html part is clear so far but how do I pass the parameter in the corresponding route area?

```


timer = """<!DOCTYPE HTML><html><head>
            <title>Timer</title>
            <body>
            <form name="timer" action="/get" target="hidden-form">
              <h2>Timer</h2>
              <table class="center" border="0">
              <tr>
              <td>Zeitzone:</td>
              </tr>
              <tr>
              <td><input type="text" size="15" name="timezone" placeholder='Posix TZ Format' value=timezone></td>
                .
                .
                .
             </body></html>"""
```

    @app.route('/timer')
    def index(request):
        #return 'Hello, world!'
        return timer, timezone="my-timezone" ,200, {'Content-Type': 'text/html'}

**Reply by miguelgrinberg** [reply]
You can use a template library such as utemplate if you want something robust, or else you can use the format method of strings. Something like this:

```python
timer = """
.
.
<td><input type="text" size="15" name="timezone" placeholder='Posix TZ Format' value="{timezone}"></td>
.
.
"""

@app.route('/timer')
def index(request):
    #return 'Hello, world!'
    return timer.format(timezone="my-timezone"), 200, {'Content-Type': 'text/html'}

**Comment by danieldownload** [comment]
Hmmm, that gives me the following error as soon as I open the timer page:

network config: ('192.168.179.xxxx'....)

ws
Traceback (most recent call last):
  File "microdot.py", line 1151, in dispatch_request
  File "ws.py", line 180, in index
KeyError: 
                alert("Ok");
                setTimeout(function(){ document

<img width="769" alt="Bildschirm­foto 2023-03-15 um 16 03 59" src="https://user-images.githubusercontent.com/26095164/225351729-5ea6a5c8-db68-4e60-831c-044cfdbc7660.png">

**Reply by miguelgrinberg** [reply]
Ah yes, this is because the `format()` function uses the `{` and `}` as template placeholders, so your JavaScript code confuses it. You will need to use utemplate instead, for which I have an integration in this package.

utemplate: https://github.com/pfalcon/utemplate
utemplate integration with Microdot: https://microdot.readthedocs.io/en/latest/extensions.html#using-the-utemplate-engine

**Reply by ghost** [reply]
Another solution with the python `format()` function would be to use doubling braces `{{` and `}}` instead of single ones `{` and `}` in the template (for css and js), and leaving single braces those containing template placeholders.

**Comment by danieldownload** [comment, marked as answer]
Sorry but I can't get it to work.

The microdot_utemplate.py refers to "from utemplate import recompile"

I cannot find a utemplate.py file on their GitHub page.

Subsequently, I can't get the method "render_template" running.

What am I missing here?

**Reply by miguelgrinberg** [reply]
You have to copy the utemplate directory to your project.

**Comment by danieldownload** [comment]
That worked, thanks a lot.
